pkg/providers/subnet/subnet.go

```diff
@@ -103,29 +111,48 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
 
 // CheckAnyPublicIPAssociations returns a bool indicating whether all referenced subnets assign public IPv4 addresses to EC2 instances created therein
 func (p *DefaultProvider) CheckAnyPublicIPAssociations(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (bool, error) {
-	subnets, err := p.List(ctx, nodeClass)
-	if err != nil {
-		return false, err
+	subnets := map[string]bool{}
+	for _, subnet := range nodeClass.Status.Subnets {
+		if subnetAssociatePublicIP, ok := p.associatePublicIPAddressCache.Get(subnet.ID); ok {
+			subnets[subnet.ID] = subnetAssociatePublicIP.(bool)
+		}
 	}
-	_, ok := lo.Find(subnets, func(s *ec2.Subnet) bool {
-		return aws.BoolValue(s.MapPublicIpOnLaunch)
-	})
-	return ok, nil
+
+	for _, v := range subnets {
+		if v {
+			return true, nil
+		}
+	}
+	return false, nil
 }
 
 // ZonalSubnetsForLaunch returns a mapping of zone to the subnet with the most available IP addresses and deducts the passed ips from the available count
 func (p *DefaultProvider) ZonalSubnetsForLaunch(ctx context.Context, nodeClass *v1beta1.EC2NodeClass, instanceTypes []*cloudprovider.InstanceType, capacityType string) (map[string]*ec2.Subnet, error) {
-	subnets, err := p.List(ctx, nodeClass)
-	if err != nil {
-		return nil, err
-	}
-	if len(subnets) == 0 {
-		return nil, fmt.Errorf("no subnets matched selector %v", nodeClass.Spec.SubnetSelectorTerms)
-	}
 	p.Lock()
 	defer p.Unlock()
-	// sort subnets in ascending order of available IP addresses and populate map with most available subnet per AZ
+
+	if len(nodeClass.Status.Subnets) == 0 {
```

Comment:
Does this being below the lock change anything?

Reply:
It won't change much, but it would be better to exit on loops sooner if they will fail this check